ToolBox/Polygon/PolygonDataQueueHandler.cs

```diff
@@ -507,105 +508,235 @@ private IEnumerable<Tick> GetCryptoTradeTicks(HistoryRequest request)
+                            {
+                                // Parse the row
+                                var row = objTick.ToObject<EquityQuoteTickResponse>();
+
+                                // Converts the results tick time to the Exchange time zone for usage in the Tick Writer LeanData.cs
+                                var utcTime = Time.UnixNanosecondTimeStampToDateTime(row.SipTimestamp);
+                                var time = GetTickTime(request.Symbol, utcTime);
+
+                                // Get the timestamp of the last tick in the results and use for pagination of the API with offset
+                                if (objTicks.Count() == _apiResultsLimit)
+                                {
+                                    offsetStart = objTicks.Last.ToObject<EquityQuoteTickResponse>().SipTimestamp;
+                                }
+                                else
+                                {
+                                    offsetStart = offsetEnd;   // Finish
+                                }
+
+                                // TODO:   This logic should be enhanced by using the corrections to flag older ticks and better quote filtering
+                                // Look for quotes.   In this case, it uses the Polygon Quote conditions mapping for conditions that are suspicious and will be filtered
+                                bool _suspiciousFlag = false;
+
+                                // Set flag if there are any trade correction indicators    //TODO:   fix these quote and trade conditions logic ... not right !  Will bad quotes have null conditions ?
+                                if ((row.Conditions != null))
+                                {
+                                    _suspiciousFlag = (PolygonQuoteMappings.isSuspiciousQuote(row.Conditions[0]));
+                                }
+
+                                string quoteConditions = "";
+                                if (row.Conditions != null)
+                                {
+                                    quoteConditions = string.Join("", row.Conditions.Select(p => String.Format(System.Globalization.CultureInfo.GetCultureInfo("en-US"), "{0:X2}", p)));
+                                }
+                                // Adds the exchange the Tick.   Prefers the bid exchange and assumes mutually exclusive bid/ask exchange IDs on quote
+                                string exchange = "";
+                                if (row.BidExchange != null)
+                                {
+                                    exchange = row.BidExchange.ToString(System.Globalization.CultureInfo.GetCultureInfo("en-US"));
+                                }
+                                else if (row.AskExchange != null)
+                                {
+                                    exchange = row.AskExchange.ToString(System.Globalization.CultureInfo.GetCultureInfo("en-US"));
```

Comment:
We should use the exchanges defined by https://github.com/QuantConnect/Lean/pull/5194/files at https://github.com/QuantConnect/Lean/blob/master/Common/Global.cs#L37